pkg/js/validator/validate-store.ts

```diff
@@ -0,0 +1,976 @@
+  return invalidStore(`invalid relation '${relation}'. Valid relations are [${relations}]`, instancePath);
+};
+
+const nonMatchingRelationType = (relation: string, user: string, values: string[], instancePath: string) => {
+  if (values.length) {
+    return invalidStore(
+      `\`${relation}\` is not a relation on \`${user}\`, and does not exist in model - valid relations are [${values}].`,
+      instancePath,
+    );
+  }
+  return invalidStore(`\`${relation}\` is not a relation on \`${user}\`, and does not exist in model.`, instancePath);
+};
+
+const invalidAssertion = (assertion: string, object: string, instancePath: string) => {
+  return invalidStore(`\`${assertion}\` is not a relationship for type \`${object}\`.`, instancePath);
+};
+
+const unidentifiedTestParam = (testParam: string, instancePath: string) => {
+  return invalidStore(
+    `\`${testParam}\` is not a recognized paramaeter for any condition defined in the model.`,
+    instancePath,
+  );
+};
+
+const undefinedTypeTuple = (user: string, instancePath: string) => {
+  return {
+    keyword: "valid_store_warning",
+    message: `${user} does not match any existing tuples; the check is still valid - but double check to ensure this is intended.`,
+    instancePath,
+  };
+};
+
+// Helpers
+
+function isValidTuple(tuple: TupleKey): boolean {
+  if (
+    !tuple.user ||
+    !isStringValue(tuple.user) ||
+    !tuple.object ||
+    !isStringValue(tuple.object) ||
+    !tuple.relation ||
+    !isStringValue(tuple.relation)
+  ) {
+    return false;
+  }
+
+  return true;
+}
+
+function mapTuple(tuple: TupleKey): RelationReference {
+  return {
+    type: tuple.user.split(":")[0],
+    relation: tuple.user.includes("#") ? tuple.user.split("#")[1] : undefined,
+    wildcard: tuple.user.includes(":*") ? {} : undefined,
+    condition: tuple.condition?.name,
+  };
+}
+
+// Validation for Types
+
+function validateTypes(tuple: TupleKey, types: string[], instancePath: string): boolean {
+  const errors = [];
+
+  const user = tuple.user.split(":")[0];
+
+  // Ensure valid type of user
+  if (!types.includes(user)) {
+    errors.push(invalidType(user, types, instancePath + "/user"));
+  }
+
+  const object = tuple.object.split(":")[0];
+
+  // Ensure valid type of object
+  if (!types.includes(object)) {
+    errors.push(invalidType(object, types, instancePath + "/object"));
+  }
+
+  // Report all errors
+  if (errors.length) {
+    validateTuple.errors?.push(...errors);
+    return false;
+  }
+  return true;
+}
+
+// Validate Relation
+function validateRelation(tuple: TupleKey, typeDefs: TypeDefinition[], instancePath: string): boolean {
+  const errors = [];
+
+  // Check if relation exists on given type
+  let doesExistOnType = false;
+  if (tuple.user.includes("#")) {
+    const user = tuple.user.split(":")[0];
+    const userRelation = tuple.user.split("#")[1];
+    for (const typeDef of typeDefs) {
```

Comment:
In a follow-up PR, let's pre-compute these. The language representation of the model should have all the relations per type precomputed so as to improve performance.

Reply:
https://github.com/openfga/language/issues/290